### DISCUSSION TITLE
How to port from HoloToolKit to MRTK?

### DISCUSSION THREAD

**Discussion by landnn** [microsoft/MixedRealityToolkit-Unity#9139] [Q&A] [answered]
Hi all, me and my team are working on getting our app originally built for HoloLens 1 onto the HoloLens 2, but we don't know how to port from HoloToolKit to MRTK. Could anyone give us some pointers? We found this [article](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/HTKToMRTKPortingGuide.html) but it doesn't provide clear instruction on how to complete the process. This is an undergraduate research project we're working on and our first experience with Unity and the HoloLens, so we don't completely understand how the two work together. We've updated the Unity project to 2019.4, but that's about all. How do we get MRTK to work with an app originally built with HoloToolKit?

**Comment by ThanosRestas** [comment, marked as answer]
You have to actually replace HoloToolKit method and interfaces with the new ones in MRTK. For example :
- [This table shows how the HTK interfaces used for anything input related looks in MRTK](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/HTKToMRTKPortingGuide.html#interface-and-event-mappings )
- [This table shows how the setup of the main camera in the scene is changed to work in the MRTK version](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/HTKToMRTKPortingGuide.html#camera)

It might seem tiresome and annoying but you have to actually go there and take it bit by bit upgrading each part of your code. The result will be worth it since you are going to follow the latest updates and discussions, thus finding help online will be easier. 

Good luck on your project most people here would have killed to have started with AR during their undergraduate studies. Starting early is the secret 👌

**Reply by landnn** [reply]
Do I have to delete any of the HoloToolKit assets? After importing MRTK I get a compiler issue saying "For Mixed Reality Toolkit input to work properly, please remove your other input module(s) and add a MixedRealityInputModule to your main camera.". However, after deleting HTK InputManager and reconfiguring MRTK the problem still persists. Also, after importing MRTK there isn't a "MixedRealityPlayspace" component in the scene like there should be. Is there another step I should be aware of? I'm sorry if these questions are awkward, I really appreciate your help.

**Reply by ThanosRestas** [reply]
Be sure to follow the guidelines for the correct setup of MRTK first, [here it is from the official documentation](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Installation.html#2-import-mrtk-packages-into-your-unity-project)

I suggest using this guide to create a new clean/empty project and start from there importing bit by bit your previous work, just to be sure.

Your **MixedRealityPlayspace** missing is solved by **Add to Scene and Configure**

![image](https://user-images.githubusercontent.com/5999342/105154001-bab61980-5b11-11eb-92fb-bb0e56eb96e2.png)

No questions are awkward, we are learning together. Take it slow from the start don't try to copy paste things/packages etc and you will be there faster. I promise. 

Focus on the porting process instead of the porting end product. I know you are eager to see it, all of us know this feeling.

**Reply by landnn** [reply]
So if I start rebuilding the project from scratch, you’re saying there’s a difference between reimporting my assets versus copying and pasting them into the new project?

**Reply by ThanosRestas** [reply]
Simple copy pasting won't suffice. There will be conflicts probably so try to make each component work separately and fix as you go.
